### PR TITLE
Feature: Subquery for SELECT

### PR DESCRIPTION
**Pull request by iRedds**
**Description**
Added ``BaseBuilder::selectSubquery(BaseBuilder $subquery, string $as)`` method

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: user_guide_src/source/changelogs/v4.2.0.rst [hunk 1]
```diff
@@ -24,6 +24,7 @@ Enhancements
     - See :ref:`content-security-policy` for details.
 - New View Decorators allow modifying the generated HTML prior to caching.
 - Added Subqueries in the FROM section. See :ref:`query-builder-from-subquery`.
+- Added Subqueries in the SELECT section.
```
Comment (lonnieezell) [inline user_guide_src/source/changelogs/v4.2.0.rst:27]: Please add link to the docs  like the surrounding changes have with theirs.

**Review by lonnieezell — CHANGES_REQUESTED** [review]
Man, I just struggled with this on a different framework for the day job and wish this would have been available there. Nice addition, just one small ask.

File: user_guide_src/source/database/query_builder.rst [hunk 2]
```diff
@@ -167,6 +167,8 @@ the resulting field.
 .. literalinclude:: query_builder/014.php
    :lines: 2-
 
+.. _query-builder-select-subquery:
```
Comment (kenjis) [inline user_guide_src/source/database/query_builder.rst:170]: @iRedds The label must  precede a section header.

`**$builder->selectSubquery()**` is not a section.

File: system/Database/BaseBuilder.php [hunk 3]
```diff
@@ -445,6 +445,22 @@ public function selectCount(string $select = '', string $alias = '')
...
 
+    /**
+     * Adds a subquery to the selection
+     *
+     * @return static
+     */
+    public function selectSubquery(BaseBuilder $subquery, string $as)
+    {
+        if (! $this->isSubquery($subquery)) {
+            throw new DatabaseException('The BaseBuilder::selectSubquery method expects a BaseBuilder instance');
```
Comment (kenjis) [inline system/Database/BaseBuilder.php:456]:
```suggestion
            throw new DatabaseException('The BaseBuilder::selectSubquery() method expects a BaseBuilder instance');
```

File: user_guide_src/source/database/query_builder/098.php [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+$subquery = $db->table('countries')->select('name')->where('id', 1);
+$builder  = $db->table('users')->select('name')->selectSubquery($subquery, 'country');
+$query = $builder->get();
+
+// Produces: SELECT `name`, (SELECT `name` FROM `countries` WHERE `id` = 1) AS `country` FROM `users`
```
Comment (kenjis) [inline user_guide_src/source/database/query_builder/098.php:7]: Please add a newline.

**Review by kenjis — DISMISSED** [review]

File: system/Database/BaseBuilder.php [hunk 5]
```diff
@@ -445,6 +445,24 @@ public function selectCount(string $select = '', string $alias = '')
         return $this->maxMinAvgSum($select, $alias, 'COUNT');
     }
 
+    /**
+     * Adds a subquery to the selection
+     *
+     * @return static
+     */
+    public function selectSubquery(BaseBuilder $subquery, string $as)
```
Comment (kenjis) [inline system/Database/BaseBuilder.php:453]:
```suggestion
    public function selectSubquery(BaseBuilder $subquery, string $as): self
```

File: system/Database/BaseBuilder.php [hunk 6]
```diff
@@ -445,6 +445,24 @@ public function selectCount(string $select = '', string $alias = '')
         return $this->maxMinAvgSum($select, $alias, 'COUNT');
     }
 
+    /**
+     * Adds a subquery to the selection
+     *
+     * @return static
```
Comment (kenjis) [inline system/Database/BaseBuilder.php:451]:
```suggestion
     * @return $this
```

Comment (iRedds) [reply system/Database/BaseBuilder.php:451]: If you specify a return type after the method, then the rector removes the @return from the docblock.
```
     /**
      * Adds a subquery to the selection
      *
-     * @return static
      */
     public function selectSubquery(BaseBuilder $subquery, string $as): self
     {
    ----------- end diff -----------

Applied rules:
 * RemoveUselessReturnTagRector

```

Comment (kenjis) [reply system/Database/BaseBuilder.php]: Okay.

**Review by kenjis — APPROVED** [review]
LGTM

**Review by paulbalandan — APPROVED** [review]

File: system/Database/BaseBuilder.php [hunk 7]
```diff
@@ -445,6 +445,22 @@ public function selectCount(string $select = '', string $alias = '')
         return $this->maxMinAvgSum($select, $alias, 'COUNT');
     }
 
+    /**
+     * Adds a subquery to the selection
+     */
+    public function selectSubquery(BaseBuilder $subquery, string $as): self
+    {
+        if (! $this->isSubquery($subquery)) {
```
Comment (samsonasik) [inline system/Database/BaseBuilder.php:453]: looking at mobile, it seems `$subquery` already type hinted, passed non `BaseBuilder` will already cause instance error so isSubquery() check no longer needed?

Comment (iRedds) [reply system/Database/BaseBuilder.php]: You're right, this check was unnecessary

**Review by samsonasik — APPROVED** [review]

**Comment by samsonasik** [comment]
@iRedds looking good, how about support both `BaseBuilder` and `Closure` with `@param BaseBuilder|Closure $subquery` with remove param type hint so the `->isSubquery()` check will make sense, as buildSubquery() support both types?

**Comment by iRedds** [comment]
Honestly, I do not know. It seems to me that using a closure as a non-last argument worsens the readability of the code.

**Comment by samsonasik** [comment]
Ok then 👍

**Comment by houjl04** [comment]
The query statement “select... for update”, how do I do that? Is there a method?

**Pull request merged**